vampyr/treebuilders/arithmetics.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <MRCPP/treebuilders/add.h>
+#include <MRCPP/treebuilders/multiply.h>
+
+namespace vampyr {
```

Comment:
These can (probably) also be bound to overloaded versions of `+`, `-`, `*`, and `**` (unary and binary)

Reply:
If you set defaults for all other parameters, that is.

Reply:
I can (will) overload the `**` and `*= c` operators.

`+= tree` and `*= tree` cannot be overloaded as the rhs is not `const` in our case (the trees might temporarily change during the operation if the grids are not directly compatible). This can probably be forcibly circumvented either on the VAMPyR or the MRCPP side, but I was not able to figure that out. The proper way to do this is probably to introduce some `mutable` parts of the `FunctionTree` for such temporary changed states, but this might get complicated.

`tree_c = tree_a + tree_b` and `tree_c = tree_a * tree_b` cannot be overloaded because `tree_c` is an `inout` argument in these operations.

Reply:
Is it possible to make Vampyr deduct the type of `tree_c`. So we can avoid the line `tree_c = vp.D3.FunctionTree(MRA)` above it?

Reply:
The problem is that `tree_c` has a state before the operation, and you would generally do the union-grid addition:
```
tree_c = vp.D3.FunctionTree(MRA)
vp.build_grid(out=tree_c, inp=tree_a)
vp.build_grid(out=tree_c, inp=tree_b)
vp.add(out=tree_c, a=1.0, inp_a=tree_a, b=1.0, inp_b=tree_b)
```
alternatively, the grid is precision driven
```
tree_c = vp.D3.FunctionTree(MRA)
vp.add(prec=epsilon, out=tree_c, a=1.0, inp_a=tree_a, b=1.0, inp_b=tree_b)
```
which is also not compatible with operator overloading

Reply:
But, we _could_ overload the operator to mean exactly the first case, which in general is what we want (at least in a vampyr setting).

Reply:
I think that would make sense, you can always chose to use `vp.add` directly if you want more control.

Reply:
Ok, this works
```
def tree_add(self, inp):
      out = vp.D1.FunctionTree(self.getMRA())
      vp.build_grid(out, self)
      vp.build_grid(out, inp)
      vp.add(-1.0, out, 1.0, self, 1.0, inp, -1)
      return out

vp.D1.FunctionTree.__add__ = tree_add

tree_c = tree_a + tree_b
```
Where would be the appropriate place to put this in vampyr? Note that the plan is to add a lot more extensions like this in the future.

Reply:
You mean things that are not pure bindings? Maybe we can split  the `vampyr` directory into `vampyr/bindings` and `vampyr/features`?

Reply:
This is the canonical way of doing it: https://github.com/pybind/scikit_build_example/tree/master/src

Reply:
Ok, this becomes a bit too involved, so I will add it to the [wishlist](#49). For this PR we only provide bindings

Reply:
So they would be in the `vampyr` folder directly. Might need some CMake massaging to copy the files in the right place when building/installing. You can bind this in the C++ code for `FunctionTree`:

```c++
.def("__add__", [](FunctionTree<D>& tree_a, FunctionTree<D>& tree_b) { 
        /* code here */
},
            py::is_operator());
```

Reply:
In general, I prefer binding in C++ code as much as possible, rather than modifying bound classes after the fact.

Reply:
`FunctionTrees` are non-copyable, but this seems to work (but I feel uncomfortable :slightly_smiling_face: )
```
        .def("__add__", [](FunctionTree<D> &rhs, FunctionTree<D> &lhs) {
            auto *out = new FunctionTree<D>(rhs.getMRA());
            mrcpp::copy_grid(*out, rhs);
            mrcpp::copy_grid(*out, lhs);
            mrcpp::add(-1.0, *out, 1.0, rhs, 1.0, lhs);
            return out;
        }, py::is_operator())
```

Reply:
I think you need to wrap them in a shared pointer and/or make sure that memory management is handed over to Python (the garbage collector can do much harm otherwise). Let's leave non-trivial overloads to the wishlist.

Reply:
Related issue: Our `FunctionTreeVector` is a `std::vector<std::tuple<double, FunctionTree<D> *>>`, and the `mrcpp::gradient` function returns such a vector. Is this an appropriate and safe way of binding it (it works)?
```
m.def("gradient",
      [] (DerivativeOperator<D> &oper, FunctionTree<D> &inp) {
          auto tmp = mrcpp::gradient(oper, inp);
          std::vector<std::unique_ptr<FunctionTree<D>>> out;
          for (int i = 0; i < tmp.size(); i++) {
              auto *tmp_i = std::get<1>(tmp[i]);
              out.push_back(std::unique_ptr<FunctionTree<D>>(tmp_i));
          }
          mrcpp::clear(tmp, false);
          return out;
      },
      "oper"_a,
      "inp"_a);
```

Is there any way of checking the resulting code for memory leaks?

Reply:
The return value is a `std::vector<std::unique_ptr<FunctionTree<D>>>`? It seems safe, but probably better with `std::shared_ptr<FunctionTree<D>>` You can run a Python script through Valgrind, though it will give you a lot of false positives from CPython too (you need a suppression file to remove those, but I don't have one at hand right now)

Reply:
hmm, with `shared_ptr` is segfaults when I access the result. will keep it as `unique`